drivers/platform/mbed/mbed_irq.cpp

```diff
@@ -0,0 +1,569 @@
+ *			EXTERNAL_INT_ID4 is triggered.
+ */
+static void mbed_ext_int_id4_callback(void)
+{
+	if (mbed_irq_callbacks.callback_ext_int_id4.callback)
+		/* Invoke the application registered callback function */
+		mbed_irq_callbacks.callback_ext_int_id4.callback(
+			mbed_irq_callbacks.callback_ext_int_id4.ctx, EXTERNAL_INT_ID4, NULL);
+}
+
+/**
+ * @brief	Mbed callback function for external interrupt ID5 event.
+ * @return	none.
+ * @note	This function is called when external IRQ event mapped to
+ *			EXTERNAL_INT_ID5 is triggered.
+ */
+static void mbed_ext_int_id5_callback(void)
+{
+	if (mbed_irq_callbacks.callback_ext_int_id5.callback)
+		/* Invoke the application registered callback function */
+		mbed_irq_callbacks.callback_ext_int_id5.callback(
+			mbed_irq_callbacks.callback_ext_int_id5.ctx, EXTERNAL_INT_ID5, NULL);
+}
+
+/**
+ * @brief	Mbed callback function for UART Rx ID1 event.
+ * @return	none.
+ * @note	This function is called when external IRQ event mapped to
+ *			UART_RX_INT_ID1 is triggered.
+ */
+static void mbed_uart_rx_id1_callback(void)
+{
+	if (mbed_irq_callbacks.callback_uart_rx_id1.callback)
+		/* Invoke the application registered callback function */
+		mbed_irq_callbacks.callback_uart_rx_id1.callback(
+			mbed_irq_callbacks.callback_uart_rx_id1.ctx, UART_RX_INT_ID1, NULL);
+}
+
+/**
+ * @brief	Mbed callback function for ticker ID event.
+ * @return	none.
+ * @note	This function is called when external IRQ event mapped to
+ *			TICKER_INT_ID is triggered.
+ */
+static void mbed_ticker_id_callback(void)
+{
+	if (mbed_irq_callbacks.callback_ticker_id.callback)
+		/* Invoke the application registered callback function */
+		mbed_irq_callbacks.callback_ticker_id.callback(
+			mbed_irq_callbacks.callback_ticker_id.ctx, TICKER_INT_ID, NULL);
+}
+
+/**
+ * @brief	Initialized the controller for the peripheral interrupts.
+ * @param	desc[in, out] - Pointer where the configured instance is stored.
+ * @param	param[in] - Configuration information for the instance.
+ * @return	SUCCESS in case of success, negative error code otherwise.
+ * @note	Supports only External, Ticker and Unbuffered UART Rx IRQs.
+ */
+int32_t mbed_irq_ctrl_init(struct irq_ctrl_desc **desc,
+			   const struct irq_init_param *param)
+{
+	mbed::Ticker *ticker;
+	InterruptIn *ext_interrupt;
+	struct irq_ctrl_desc *irq_desc;
+	struct mbed_irq_desc *mbed_irq_desc;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	irq_desc = (struct irq_ctrl_desc *)calloc(1, sizeof(*irq_desc));
+	if (!irq_desc)
+		return -ENOMEM;
+
+	mbed_irq_desc = (struct mbed_irq_desc *)calloc(1, sizeof(*mbed_irq_desc));
+	if (!mbed_irq_desc)
+		goto err_mbed_irq_desc;
+
+	irq_desc->irq_ctrl_id = param->irq_ctrl_id;
+	mbed_irq_desc->int_mode = ((struct mbed_irq_init_param *)
+				   param->extra)->int_mode;
+	mbed_irq_desc->ext_int_pin = ((struct mbed_irq_init_param *)
+				      param->extra)->ext_int_pin;
+
+	switch (param->irq_ctrl_id) {
```

Comment:
Here, the `irq_ctrl_id` member stands for the ID of the interrupt controller, not the ID of the interrupt. This is so that we can have multiple interrupt controllers in the same system, possibly cascading into each other. We should just save this in the `irq_desc` and do this `switch case` later, in the register interrupt method. Is it possible to do this?

Reply:
Sorry not sure if I got your comment correctly, so just clarifying it.

Do you mean to say that ```irq_id``` passed as an argument to other functions in this module is not the same as ```irq_ctrl_id```? That means the ```irq_id enum``` value from ```mbed_irq.h``` file will be passed as argument to ```irq_id``` parameter for all other functions except this init function. The code under this ```switch()``` case would be then part of ```mbed_irq_register_callback()``` function.  
If this the case, then it make sense. I was also wondering why ```irq_id``` is passed again to all the functions if it is already passed to init function via irq init structure, but both seems to be different. 

Please confirm my understanding.

Reply:
Yes, you are right. `irq_ctrl_id` is the ID of the interrupt controller that is used, and is set in the init function, and `irq_id` is the ID of the interrupt within the interrupt controller.

Reply:
Thanks for the confirmation. I will implement the changes accordingly.